### DISCUSSION TITLE
Are back-ups created for storage?

### DISCUSSION THREAD

**Discussion by nthouliss** [supabase/supabase#6755] [Questions] [answered]
I know that the database is backed-up on the paid tiers, however does this extend to Storage? I know that the documentation says this:

```
Database backups do NOT include objects stored via the Storage API, as the database only includes metadata about these objects. Restoring an old backup will not restore objects that have been deleted since then.
```

I suppose I'm a little confused in the wording in the last sentence and I just want to be sure, are the buckets with their objects also get backed up regularly? If they don't is there any mechanism that I can use to do this?

Cheers :)

**Comment by inian** [comment, marked as answer]
Hi @nthouliss, the objects are stored in AWS S3 which has extremely high durability (99.999999999%). So there is very little chance that your objects are lost. 

However, the objects are not part of the daily backup. So if you delete an object from storage, it is permanently deleted. Even if you restore a backup from before the object was deleted, it won't restore the deleted object. 

We have a feature where you can bring your own storage bucket to our hosted platform. With that in place, you can snapshot the objects in Storage and back them up based on your requirements.

**Reply by nthouliss** [reply]
> the objects are stored in AWS S3 which has extremely high durability (99.999999999%). So there is very little chance that your objects are lost.
 
It's less that I'm worried about system failure and more that I want to be assured that I'm protecting against myself.

> We have a feature where you can bring your own storage bucket to our hosted platform. With that in place, you can snapshot the objects in Storage and back them up based on your requirements.

Is there any official documentation around this? I can't seem to find it. I have seen mentions of it around GitHub issues/discussions but I wasn't aware that this was released or how to go about doing it.

**Reply by inian** [reply]
Hey @nthouliss, ah that was a typo on my side. We haven't launched that yet. You can track our progress [here](https://github.com/supabase/storage-api/issues/27)

**Reply by nthouliss** [reply]
No worries, are there any conversations about allowing object versioning on the hosted version of the site? This would really help in the idea of protecting users against themselves.

**Reply by eqoram** [reply]
This is definitely needed

**Reply by flobo79** [reply]
I would love to see a backup solution for Supasebase storage too, especially to keep the key proposition as a paas with batteries included. happy to pay extra for it too.

**Reply by gonpeche** [reply]
Hey all,
Is there a backup solution already for the storage system?

**Comment by gruckion** [comment]
Er... can you please enable the versioning feature on the S3 buckets please and expose it via your API. Thank you <3

I am assuming your using AWS S3 

https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html

@nthouliss

**Comment by supertj** [comment]
@gonpeche asked this in February and it's still open — and since this thread still ranks for the question, here's where things stand in 2026: unchanged. Supabase's automated backups cover the **database only**. Storage files are not in any backup you can restore. Worse, if you *do* restore a database backup, the `storage.objects` metadata rows come back and point at files the restore did not bring back, so your app looks intact until someone opens an upload.

What you can do today with standard tools:

**1. Turn on the S3 protocol for your project's Storage.** Dashboard → **Storage → Settings → S3 Access Keys** → enable the S3 connection and generate an access key pair. The endpoint and region are shown on the same page. (These keys bypass RLS — keep them server-side only.)

**2. Sync every bucket to storage you own** — e.g. Cloudflare R2, Backblaze B2, or your own S3 — with rclone:

```ini
# ~/.config/rclone/rclone.conf
[supabase]
type = s3
provider = Other
access_key_id = <SUPABASE_S3_ACCESS_KEY_ID>
secret_access_key = <SUPABASE_S3_SECRET>
endpoint = https://<project-ref>.storage.supabase.co/storage/v1/s3
region = <project-region>

[r2]
type = s3
provider = Cloudflare
access_key_id = <R2_ACCESS_KEY_ID>
secret_access_key = <R2_SECRET>
endpoint = https://<account-id>.r2.cloudflarestorage.com
```

```bash
rclone sync supabase:<bucket> r2:my-backups/storage/<bucket> --checksum
```

If rclone throws signature errors against the Supabase endpoint (some versions did — see discussion #22200), the AWS CLI is a reliable fallback:

```bash
aws s3 sync s3://<bucket> ./storage-backup/<bucket> \
  --endpoint-url "https://<project-ref>.storage.supabase.co/storage/v1/s3" \
  --region <project-region>
```

**3. Schedule it, and dump the database in the same run** (cron or a GitHub Actions workflow), so the DB snapshot and the file snapshot describe the same moment — otherwise a restore gives you metadata rows and files that disagree with each other.

**4. Mind egress.** Downloading your files counts as *uncached egress* on your Supabase bill: $0.09/GB beyond your plan's allowance (Free: 5 GB/month, enforced through a fair-use flow of notice → grace period → restrictions, so recurring large backups aren't viable there; Pro: 250 GB included). For big buckets, weekly beats daily.

I ended up building an open-source CLI for exactly this — disclosure: I'm the author. `npx backupdrill backup` captures the Postgres dump and (given your project's Storage S3 keys) the Storage files together into your own bucket with a SHA-256 manifest, and `backupdrill drill` restores the snapshot into a throwaway Docker Postgres and checksums the files to prove the backup actually restores: https://github.com/backupdrill/cli (MIT, comes with a GitHub Actions template for scheduling).

Step-by-step version of the manual approach above: https://backupdrill.com/guides/supabase-storage-backup
